### PR TITLE
Wip/6.0 cascade 

### PR DESCRIPTION
**Pull request by dreab8**
re-enabling few tests for cascade emerged few other non related issues, this PR contains the fix for all these issues:

- issue with circularity detection
- NPE when a fetched association is null 
- EntityFetchDelayedInitializer check if the entity instance has been already loaded
- EntityInitializer and DelayedCollectoinAssembler see if we have an existing entry in the PC for the EntityKey 
-  Issue with initialization of empty collections 
- issue with SimpleForeignKeyDescriptor#generateJoinPredicate() using wrong columns for selfreferenced associations
- issue with SimpleForeignKeyDescriptor DomainResult creation

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/SingularAssociationAttributeMapping.java [hunk 1]
```diff
@@ -238,33 +208,35 @@ public Fetch resolveCircularFetch(
...
-					if ( targetColumns.equals( identifierColumnNames[0] ) ) {
-						return createBiDirectionalFetch( fetchablePath, fetchParent );
-					}
+						if ( identifierColumnNames.length > 1 ) {
+							throw new NotYetImplementedFor6Exception(
+									"Support for composite foreign -keys not yet 	implemented" );
+						}
+						if ( targetColumns.equals( identifierColumnNames[0] ) ) {
+							return createBiDirectionalFetch( fetchablePath, fetchParent );
+						}
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/SingularAssociationAttributeMapping.java]: seems there is some indentation alignment issue. not a big deal for sure.

File: hibernate-core/src/main/java/org/hibernate/sql/results/graph/collection/internal/DelayedCollectionInitializer.java [hunk 2]
```diff
@@ -16,98 +16,132 @@
...
-				owner -> collectionInstance.setOwner( owner )
-		);
+		final Object parentKey = parentAccess.getParentKey();
+		if ( parentKey != null ) {
+			collectionKey = new CollectionKey(
+					fetchedMapping.getCollectionDescriptor(),
+					parentKey
+			);
+
+			parentAccess.registerResolutionListener( owner -> collectionInstance.setOwner( owner ) );
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/sql/results/graph/collection/internal/DelayedCollectionInitializer.java]: seems we can rely on method reference to simply the above as
```
parentAccess.registerResolutionListener( collectionInstance::setOwner );
```

File: hibernate-core/src/test/java/org/hibernate/orm/test/cascade/RefreshTest.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+ * @author Steve Ebersole
+ */
+@DomainModel(
+		xmlMappings = {
+				"org/hibernate/orm/test/cascade/Job.hbm.xml",
+				"org/hibernate/orm/test/cascade/JobBatch.hbm.xml"
+		}
+)
+@SessionFactory
+@FailureExpected( reason = "This should be fixed by Nathan PR")
```
Comment (dreab8) [inline hibernate-core/src/test/java/org/hibernate/orm/test/cascade/RefreshTest.java]: This is the PR https://github.com/hibernate/hibernate-orm/pull/3250 fixing this test

Comment (NathanQingyangXu) [reply hibernate-core/src/test/java/org/hibernate/orm/test/cascade/RefreshTest.java]: thanks for mentioning me. I am moved by my name showing up here, :).

**Comment by dreab8** [comment]
@sebersole thanks for the review, I agree with you, it works but I'm not a huge fan too.

**Comment by dreab8** [comment]
Add a fix for loading empty array collection and merged.
Thanks @sebersole and @NathanQingyangXu

**Pull request closed without merging**